src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -634,11 +635,31 @@ private void RenderInExistingBatch(RenderQueueEntry renderQueueEntry)
                 var disposeComponentId = _batchBuilder.ComponentDisposalQueue.Dequeue();
                 var disposeComponentState = GetRequiredComponentState(disposeComponentId);
                 Log.DisposingComponent(_logger, disposeComponentState);
-                if (!disposeComponentState.TryDisposeInBatch(_batchBuilder, out var exception))
+                if (!disposeComponentState.RequiresAsyncDisposal())
+                {
+                    if (!disposeComponentState.TryDisposeInBatch(_batchBuilder, out var exception))
+                    {
+                        exceptions ??= new List<Exception>();
+                        exceptions.Add(exception);
+                    }
+                }
+                else
                 {
-                    exceptions ??= new List<Exception>();
-                    exceptions.Add(exception);
+                    var result = disposeComponentState.DisposeInBatchAsync(_batchBuilder);
+                    if (result.IsCompleted)
+                    {
+                        if (result.IsFaulted)
```

Comment:
I'm trying to check for sure whether the `IsFaulted` flag will be true for cancelled tasks. It's hard to find out because [the docs](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.isfaulted?view=netcore-3.1) for `IsFaulted` don't say whether cancellation would be counted as an exception.

However we do have [this line of code](https://github.com/dotnet/aspnetcore/blob/master/src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs#L240) which seems to suggest that "is canceled" does *not* imply "is faulted" (otherwise they could just check for `IsFaulted` and wouldn't need the `|| IsCanceled` clause). So either that line of code has some redundant check in it, or the logic on L651 here would be ignoring cancelled tasks still.

Reply:
My guess is that your code here is correct and HttpConnectionDispatcher is wrong (as in, contains a redundant check), but would like you to confirm :)

Reply:
The checks are different

>// Return true if canceled bit is set and faulted bit is not set

https://source.dot.net/#System.Private.CoreLib/Task.cs,1290

Reply:
Ah, that's clearer than the docs. Thanks!

So it seems we need to check `if (!result.IsCompletedSuccessfully)` instead of `IsFaulted` here to avoid ignoring cancelled tasks.

Reply:
From what I can tell on my unit tests, IsFaulted is enough, I haven't read the thing brennan pointed out, but I'll change the check.

Reply:
done